mapie/calibration.py

```diff
@@ -0,0 +1,559 @@
+            Training data.
+        y : ArrayLike of shape (n_samples,)
+            Training labels.
+        sample_weight : Optional[ArrayLike] of shape (n_samples,)
+            Sample weights for fitting the out-of-fold models.
+            If ``None``, then samples are equally weighted.
+            If some weights are null,
+            their corresponding observations are removed
+            before the fitting process and hence have no residuals.
+            If weights are non-uniform, residuals are still uniformly weighted.
+            Note that the sample weight defined are only for the training, not
+            for the calibration procedure.
+            By default ``None``.
+        calibrator : RegressorMixin
+            Calibrator to fit.
+
+        Returns
+        -------
+        Dict[int, RegressorMixin]
+            Dictionnary of fitted calibrators.
+        """
+        X, y = indexable(X, y)
+        y = _check_y(y)
+        max_prob, y_pred = self._get_labels(X)
+        calibrators = {}
+        for item in np.unique(y_pred):
+            calibrator_ = self._fit_calibrator(
+                item,
+                calibrator,
+                y,
+                max_prob,
+                y_pred,
+                sample_weight,
+            )
+            calibrators[item] = calibrator_
+        return calibrators
+
+    def _pred_proba_calib(
+        self,
+        idx: int,
+        item: int,
+        calibrated_values: NDArray,
+        max_prob: NDArray,
+        y_pred: NDArray,
+        calibrators: Dict[int, RegressorMixin],
+    ) -> NDArray:
+        """
+        Using the predicted scores, we calibrate the maximum score with the
+        specifically fitted calibrator. Note that if there is no calibrator
+        for a the specific class, then we simply output the not calibrated
+        values.
+
+        Note that if the calibrated score prediction is 0, there would be an
+        issue when finding the class it belongs to. We set it equal to
+        {np.finfo(np.float64).eps} as this would likely not have large impact
+        on the calibration scores, yet, would set the maximum to the correct
+        label.
+
+
+        Parameters
+        ----------
+        item : int
+            The defined subset to be calibrated.
+        calibrated_values : NDArray
+            Array of calibrated values.
+        max_prob : NDArray of shape (n_samples,)
+            Values to be calibrated.
+        y_pred : NDArray of shape (n_samples,)
+            Indicator of the values to be calibrated.
+        calibrators : Dict[RegressorMixin] of length n_classes
+            Dictionary of all the previously fitted calibrators.
+
+        Returns
+        -------
+        NDArray
+            Updated calibrated values from the predictions of the calibrators.
+
+        Raises
+        ------
+        Warnings
+            If there has not been a fitted calibrator for the specific class.
```

Comment:
Yes, it could be possible that a certain class has not been seen in the calibration sample and therefore there does not exist a fitted calibrator for it.